src/DotNetty.Codecs.Mqtt/MqttDecoder.cs

```diff
@@ -447,56 +369,36 @@ void DecodePacketIdVariableHeader(IByteBuffer buffer, PacketWithId packet, bool
 
             if (expectPayload)
             {
-                this.Checkpoint(ParseState.Payload);
-
                 this.DecodePayload(buffer); // fall-through
             }
+            else if (this.remainingLength > 0)
+            {
+                throw new DecoderException(string.Format("Remaining Length is {0} more than expected.", this.remainingLength));
+            }
         }
 
         BufferReadResult<int> DecodePacketId(IByteBuffer buffer, int currentRemainingLength)
         {
-            if (currentRemainingLength < 2)
-            {
-                throw new DecoderException(string.Format("Remaining Length is not big enough to accomodate at least 2 bytes (for packet identifier). Available value: {0}",
-                    this.remainingLength));
-            }
-
-            BufferReadResult<int> packetIdResult = this.DecodeShort(buffer);
-            if (packetIdResult.IsSuccessful)
-            {
-                Util.ValidatePacketId(packetIdResult.Value);
-            }
+            BufferReadResult<int> packetIdResult = this.DecodeShort(buffer, currentRemainingLength);
 
             return packetIdResult;
         }
 
-        void DecodeConnectPayload(IByteBuffer buffer, bool hasWill, bool hasUsername, bool hasPassword)
+        void DecodeConnectPayload(IByteBuffer buffer, bool hasWill, bool hasUsername, bool hasPassword, int bytesConsumed)
         {
             var packet = (ConnectPacket)this.currentPacket;
 
-            BufferReadResult<string> clientIdResult = this.DecodeString(buffer, this.remainingLength);
-            if (!clientIdResult.IsSuccessful)
-            {
-                return;
-            }
+            BufferReadResult<string> clientIdResult = this.DecodeString(buffer, this.remainingLength - bytesConsumed);
             string clientId = clientIdResult.Value;
             Util.ValidateClientId(clientId);
             packet.ClientId = clientId;
-            int bytesConsumed = clientIdResult.BytesConsumed;
+            bytesConsumed += clientIdResult.BytesConsumed;
 
             if (hasWill)
             {
                 BufferReadResult<string> willTopicResult = this.DecodeString(buffer, this.remainingLength - bytesConsumed);
-                if (!willTopicResult.IsSuccessful)
-                {
-                    return;
-                }
-                packet.WillTopic = willTopicResult.Value;
-                BufferReadResult<int> willMessageLengthResult = this.DecodeShort(buffer);
-                if (!willMessageLengthResult.IsSuccessful)
-                {
-                    return;
-                }
+                packet.WillTopicName = willTopicResult.Value;
+                BufferReadResult<int> willMessageLengthResult = this.DecodeShort(buffer, this.remainingLength - bytesConsumed);
                 int willMessageLength = willMessageLengthResult.Value;
                 if (!buffer.IsReadable(willMessageLength))
```

Comment:
remove RequestReplay() in this entire method. In fact, any RequestReplay() after the ensure readable is pointless